core/insertion_marker_manager.ts

```diff
@@ -460,150 +447,135 @@ export class InsertionMarkerManager {
+  private maybeHidePreview(newCandidate: CandidateConnection|null) {
     // If there's no new preview, remove the old one but don't bother deleting
     // it. We might need it later, and this saves disposing of it and recreating
     // it.
-    if (!candidate.closest) {
-      this.hidePreview_();
+    if (!newCandidate) {
+      this.hidePreview();
     } else {
-      // If there's a new preview and there was an preview before, and either
-      // connection has changed, remove the old preview.
-      const hadPreview = this.closestConnection_ && this.localConnection_;
-      const closestChanged = this.closestConnection_ !== candidate.closest;
-      const localChanged = this.localConnection_ !== candidate.local;
-
-      // Also hide if we had a preview before but now we're going to delete
-      // instead.
-      if (hadPreview &&
-          (closestChanged || localChanged || this.wouldDeleteBlock_)) {
-        this.hidePreview_();
+      if (this.activeCandidate) {
+        const closestChanged =
+            this.activeCandidate.closest !== newCandidate.closest;
+        const localChanged = this.activeCandidate.local !== newCandidate.local;
+
+        // If there's a new preview and there was an preview before, and either
+        // connection has changed, remove the old preview.
+        // Also hide if we had a preview before but now we're going to delete
+        // instead.
+        if ((closestChanged || localChanged || this.wouldDeleteBlockInternal)) {
+          this.hidePreview();
+        }
       }
     }
 
     // Either way, clear out old state.
-    this.markerConnection_ = null;
-    this.closestConnection_ = null;
-    this.localConnection_ = null;
+    this.markerConnection = null;
+    this.activeCandidate = null;
   }
 
   /**
    * A preview should be hidden.  This function figures out if it is a block
    *  highlight or an insertion marker, and hides the appropriate one.
    */
-  private hidePreview_() {
-    if (this.closestConnection_ && this.closestConnection_.targetBlock() &&
-        this.workspace_.getRenderer().shouldHighlightConnection(
-            this.closestConnection_)) {
-      this.closestConnection_.unhighlight();
-    }
-    if (this.fadedBlock_) {
-      this.hideReplacementFade_();
-    } else if (this.highlightedBlock_) {
-      this.hideInsertionInputOutline_();
-    } else if (this.markerConnection_) {
-      this.hideInsertionMarker_();
+  private hidePreview() {
+    const closest = this.activeCandidate?.closest;
+    if (closest && closest.targetBlock() &&
+        this.workspace.getRenderer().shouldHighlightConnection(closest)) {
+      closest.unhighlight();
+    }
+    if (this.fadedBlock) {
+      this.hideReplacementFade(this.fadedBlock);
+    } else if (this.highlightedBlock) {
+      this.hideInsertionInputOutline(this.highlightedBlock);
+    } else if (this.markerConnection) {
+      this.hideInsertionMarker(this.markerConnection);
```

Comment:
Calling unconditionally seems worse to me because they *are* mutually exclusive, and I think it should be clear which of the actions is actually going to happen.